src/sds.c

```diff
@@ -192,6 +192,25 @@ sds sdsdup(const sds s) {
     return sdsnewlen(s, sdslen(s));
 }
 
+/*
+ * This method returns the minimum amount of bytes required to store the sds (header + data + NULL terminator).
+ */
+static inline size_t sdsminlen(sds s) {
+    return sdslen(s) + sdsHdrSize(s[-1]) + 1;
+}
+
+/* This method copies the sds `s` into `buf` which is the target character buffer. */
+size_t sdscopytobuffer(unsigned char *buf, size_t buf_len, sds s, uint8_t *hdr_size) {
+    size_t required_keylen = sdsminlen(s);
+    if (buf == NULL) {
+        return required_keylen;
+    }
+    assert(buf_len >= required_keylen);
+    memcpy(buf, sdsAllocPtr(s), required_keylen);
```

Comment:
An sds `s` is a pointer to where the string content starts, so it can be used as a C string. It does not point to the start of the allocation. The header is store before the char data in the same allocation, i.e. at `s[-1]`, `s[-2]`, etc. The header is backwards-encoded in some way, so the byte at `s[-1]` says which kind of header it is and how large the header is.

```
       sds-----.
               |
  allocation   v
  +-----------+----------------------+
  | hdr       |string contents    \0 |
  +-----------+----------------------+
```

When we store this embedded, we want to be able to restore the sds pointer, so we store the size of the header size in the first. When we restore the sds pointer, we can find it using S (the offset to the string contents).

```  
+-+-----------+----------------------+
|S| hdr       |string contents    \0 |
+-+-----------+----------------------+
```

Reply:
You mean the callers of this function can call `sdsHdrSize(s[-1])` themselves? I'm not sure if it's public.

Reply:
But we memcpy sdsAllocPtr and not sds, so you copy with the sds header, right?

Reply:
Yea, right agree. I am not so supportive of the desire to keep something as sds when it's not, but that's already in my overall comment, maybe I'll elaborate more there. Thanks

Reply:
`key` is being used throughout the engine as a `sds` so changing that would be even more touchpoint. We could dynamically build it on the `dictGetKey` call but that would be additional penalty. Hence, storing it as a sds made the most sense.

Reply:
I have the same question, the low 3bit in sdshdr.flag have told us the header length, and exposing function `sdsHdrSize` will not lead to coupling between dict.c and sds.c, I don't see any harm in it.
```
                                                            
 sdshdr               sds                                   
   │                  │                                     
   │                  │                                     
   ▼─────┬─────┬─────┌▼───────────────────────────────┬──┐  
   │ len │ alloc     │                                │  │  
   │     │     │flags│                                │\0│  
   └─────┼─────┼─────└────────────────────────────────┴──┘  
                                                                                                                                                                                            
```